### PR TITLE
misc: Ensure cypress tab is active before any command runs

### PR DESCRIPTION
**Pull request by chrisbreiding**
### Additional details

Ensures the main Cypress tab is active (i.e. the tab on top and shown by the browser) before any command is run and at the end of a test. This is in anticipation of support for testing multiple tabs/windows with the future @cypress/puppeteer plugin. It's only implemented in Chromium for now, as that's what the puppeteer plugin will support. It's not necessary for Electron because it doesn't have tabs.

It uses the extension to accomplish activating the tab, since it's the only way to do it without the browser stealing focus from other apps (like CDP `Page.bringToFront` would do).

### Steps to test

Run any test in open mode and open a new tab while the test is running. You'll see the active tab switch back to the Cypress tab before the next command starts running. It will also switch back at the end of a test, including if a command fails.

### PR Tasks

- [x] Have tests been added/updated?
- [N/A] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [N/A] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/driver/src/cypress/browser.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ const _isBrowser = (browser, matcher, errPrefix) => {
   let exclusive = false
 
   const matchWithExclusion = (objValue, srcValue) => {
-    if (srcValue.startsWith('!')) {
+    if (_.isString(srcValue) && srcValue.startsWith('!')) {
```
Comment (chrisbreiding) [inline packages/driver/src/cypress/browser.ts:10]: This is a bit of a drive-by fix that I needed for the browser check. I didn't see any open issues about this, but it does seem like it wasn't working as documented, where `Cypress.isBrowser({ isHeadless: true })` would fail with "startsWith is not a function".

**Review by cacieprins — APPROVED** [review]

File: packages/extension/app/newtab.html [hunk 2]
```diff
@@ -13,11 +13,11 @@ <h1>Cypress is currently automating this browser.</h1>
     <p>Opening new tabs may interfere with tests and cause failures.</p>
     <p>Please note:</p>
     <ul>
-      <li>Any opened tabs will be closed when Cypress is stopped.</li>
+      <li>Any opened tabs will be closed between tests.</li>
       <li>Tests currently running may fail while another tab has focus.</li>
```
Comment (mschile) [inline packages/extension/app/newtab.html:17]: Is this still true if we set focus before each command? Or it could still fail if the user changed focus during the command.

File: packages/driver/src/util/tab_activation.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+function activateMainTab () {
+  // Don't need to activate the main tab if it already has focus
+  if (document.hasFocus()) return
+
+  return new Promise<void>((resolve) => {
+    const url = `${window.location.origin}${window.location.pathname}`
+
+    // This sends a message on the window that the extension content script
+    // listens for in order to carry out activating the main tab
+    window.postMessage({ message: 'extension:activate:main:tab', url }, '*')
```
Comment (mschile) [inline packages/driver/src/util/tab_activation.ts]: Do we want to scope these messages to Cypress in anyway?

```suggestion
    window.postMessage({ message: 'cypress:extension:activate:main:tab', url }, '*')
```

File: packages/extension/app/v3/service-worker.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    // this brings the main Cypress tab to the front of any other tabs
+    // without Chrome stealing focus from other running apps
+    await chrome.tabs.update(cypressTab.id, { active: true })
+  } catch (err) {
+    // ignore the error but log it. these logs only appear if you inspect
+    // the service worker, so it won't clutter up the console for users
+
+    // eslint-disable-next-line no-console
+    console.log('Activating main tab errored:', err)
```
Comment (mschile) [inline packages/extension/app/v3/service-worker.js]:
```suggestion
    console.log('Activating main Cypress tab errored:', err)
```

File: packages/extension/app/newtab.html (hunk 2 above)
Comment (chrisbreiding) [reply packages/extension/app/newtab.html:17]: We've made it less likely a user can effectively use a new tab while a test is running, but I think it's still true that if they do open one in the middle of a long-running command, it could interfere with that command's execution.

File: packages/driver/src/util/tab_activation.ts (hunk 3 above)
Comment (chrisbreiding) [reply packages/driver/src/util/tab_activation.ts]: Addressed in https://github.com/cypress-io/cypress/pull/28334/commits/48468486ba9da3321128140ec9109a4b6f0f1844

File: packages/extension/app/v3/service-worker.js (hunk 4 above)
Comment (chrisbreiding) [reply packages/extension/app/v3/service-worker.js]: Addressed in https://github.com/cypress-io/cypress/pull/28334/commits/48468486ba9da3321128140ec9109a4b6f0f1844

File: packages/driver/src/util/tab_activation.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+    // This sends a message on the window that the extension content script
+    // listens for in order to carry out activating the main tab
+    window.postMessage({ message: 'cypress:extension:activate:main:tab', url }, '*')
+
+    function onMessage ({ data, source }) {
+      // only accept messages from ourself
+      if (source !== window) return
+
+      if (data.message === 'main:tab:activated') {
```
Comment (mschile) [inline packages/driver/src/util/tab_activation.ts]: Also scope this one?
```suggestion
      if (data.message === 'cypress:main:tab:activated') {
```

Comment (chrisbreiding) [reply packages/driver/src/util/tab_activation.ts]: Good idea. Addressed in https://github.com/cypress-io/cypress/pull/28334/commits/639b6d01036d76179115f06434f6678b152f1c62

File: packages/extension/app/v3/service-worker.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+// here we connect to the content script, which has access to the web page
+// running Cypress, but not the extension API
+chrome.runtime.onConnect.addListener((port) => {
+  port.onMessage.addListener(async ({ message, url }) => {
+    if (message === 'activate:main:tab') {
+      await activateMainTab(url)
+
+      // send an ack back to let the content script know we successfully
+      // activated the main tab
+      port.postMessage({ message: 'main:tab:activated' })
```
Comment (mschile) [inline packages/extension/app/v3/service-worker.js:42]: Scope the `activate:main:tab` and `main:tab:activated` messages as well?

Comment (chrisbreiding) [reply packages/extension/app/v3/service-worker.js:42]: These ones are only passed within the extension, so they're already scoped in that manner. There's no risk of any event name collisions or unexpected messages coming through. The other ones are on the top window object, where that risk does exist, so it makes sense to scope the event names themselves.

**Review by mschile — APPROVED** [review]
